### PR TITLE
Quizki registration login

### PR DESCRIPTION
**Pull request by FrauDeborah**
Johnathan,

Getting closer to completion for Registration and Login. Turns out that asyncValidators do not use .subscribe(). Once I removed all the subscribes and reformat form control, like magic the validators worked. Regardless, I learned lots about Observables and Promises.

I will continue working over the weekend and push again before Sunday evening.

1) Would you please look at the apiService.get, I'm not sure that the get is implemented correctly. The apiService.get needs an Authorization so I passed username and password, and let the service setHttpHeaders Authorization for the get. Typescript would not let me overload getHttpHeaders, so I used setHttpHeaders since I was passing information.

2) Current outstanding issues include: 

- **CORS again**: The apiService.post for registerUser (/api/users/)has another CORS issue with the No 'Access-Control-Allow-Origin' I need to trouble shoot the issue.
- **Login Failure**: The login needs to catch login failures, then send back to the login form for the user to make corrections.  I'll need to review how to do that in Angular, once I understand what the backend returns for /api/verifyCredentials failure.
- **Popup Dialog**: Still need to convert to popup dialog.
- Email **Format:** the email form control still allows emails like username@gmail, rather than requiring a ".abc" Note: I backed out the pattern checking for the ".abc" because it wouldn't change invalid to valid after correction and I wanted to move forward for a change.

Let me know what if the apiService changes are correct. 

Thank you,

Deborah

### CODE REVIEW

File: src/app/_services/api.service.ts [hunk 1]
```diff
@@ -20,6 +20,24 @@ export class ApiService {
   	return rtn;
   }
 
+  // http.get needs an Authorization in header
+  setHttpHeaders(username:string, password: string) {
```
Comment (haxwell) [inline src/app/_services/api.service.ts]: I would just call this `getHttpHeadersWithUsernamePassword`.... or `getHttpHeadersWithBase64Authorization`.. or something like that..

**Comment by haxwell** [comment]
Deborah, Thank you Thank you for your work on this.. much appreciated.. 

I have updated the backend so that we will never ever (hopefully) see the CORS monster again.

Other than that, looks like you have a good idea where you need to go.. keep on coding!

I'll review it when you're done. #goodWork

**Comment by FrauDeborah** [comment]
@haxwell 

I pushed my changes because it is Sunday evening. Not much changed since Friday, just a few minor changes because I backed out some code. Still can't login new users, created a backend issue#33.  

The npm test showed 2 errors. I left the errors for now until issue#33 is resolved.

Deborah

**Comment by FrauDeborah** [comment]
@haxwell Johnathan,

1) **Methods Sample for Feedback**: Today at the StandUp meeting you asked me to write methods rather than use logic snippets in the password form control on the login form.  At this point, I changed only the password form control to use methods and variables. I would like your feedback before I modify the rest of the login form group, and modify the register form group. 

Personally, I think the form is less intuitive now (likely, because I'm accustom to seeing the logic snippets embedded in the html). Yet,  I can see how we would gain reusable code if I generalize the methods, and move to a service with methods for synchronous form controls. 

Any feedback on the methods and revised html for the password form control is appreciated.

2) **Password Show/Hide broke**: By the way, the original problem I mentioned last night in the Slack still exists...  what once worked is now broke ... i.e., after showing a password as text, it won't hide the password now. It tries because I can see a flash when clicking the icon, but something is stopping it. I'll need to research further.
 
3) **Login Error**:  When either username or password is intentionally entered incorrectly, the backend sends a 401 with the '"error":{"isTrusted":true}' which is good to know, yet, not very actionable from the users perspective. I did set the form group invalid and asked the user to re-enter the username and password.

I could submit another issue to be worked later. because it's not a show stopper. Yet, it's not nice to leave a the 401 notice and the CORS error in the console. Perhaps I can somehow capture and prevent  the 401 notice and CORS errors from displaying in the console. The entire (error) captured is error = {"headers":{"normalizedNames":{},"lazyUpdate":null,"headers":{}},"status":0,"statusText":"Unknown Error","url":null,"ok":false,"name":"HttpErrorResponse","message":"Http failure response for (unknown url): 0 Unknown Error","error":{"isTrusted":true}} 

Any advise?


Thank you.

**Comment by haxwell** [comment]
Hi @FrauDeborah .. are you still waiting for feedback from me on this, or did we already address it in Slack? Thanks.

**Comment by FrauDeborah** [comment]
Johnathan

We addressed in slack. I’ll work on it later today and tomorrow.

Thank you


Regards,

Deborah 


On Nov 3, 2018, at 10:16 PM, Johnathan James <notifications@github.com> wrote:

Hi @FrauDeborah .. are you still waiting for feedback from me on this, or did we already address it in Slack? Thanks.

**Pull request merged**